src/qutip_qip/circuit/mat_renderer.py

```diff
@@ -0,0 +1,1025 @@
+            wire_up = plt.Line2D(
+                [0, max_len],
+                [
+                    (i + self._qwires) * self.style.wire_sep,
+                    (i + self._qwires) * self.style.wire_sep,
+                ],
+                lw=1,
+                color=self.style.wire_color,
+                zorder=self._zorder["wire"],
+            )
+            wire_down = plt.Line2D(
+                [0, max_len],
+                [
+                    (i + self._qwires) * self.style.wire_sep,
+                    (i + self._qwires) * self.style.wire_sep,
+                ],
+                lw=1,
+                color=self.style.wire_color,
+                zorder=self._zorder["wire"],
+            )
+            self._ax.add_line(wire_up)
+            self._ax.add_line(wire_down)
+
+    def _add_wire_labels(self) -> None:
+        """
+        Adds the wire labels to the circuit.
+        """
+
+        if self.style.wire_label is None:
+            default_labels = [f"$c_{{{i}}}$" for i in range(self._cwires)] + [
+                f"$q_{{{i}}}$" for i in range(self._qwires)
+            ]
+            self.style.wire_label = default_labels
+
+        self.max_label_width = max(
+            [
+                self._get_text_width(
+                    label,
+                    self.style.fontsize,
+                    "normal",
+                    "monospace",
+                    "normal",
+                )
+                for label in self.style.wire_label
+            ]
+        )
+
+        for i, label in enumerate(self.style.wire_label):
+            wire_label = plt.Text(
+                -self.style.label_pad,
+                i * self.style.wire_sep,
+                label,
+                fontsize=self.style.fontsize,
+                verticalalignment="center",
+                horizontalalignment="right",
+                zorder=self._zorder["wire_label"],
+                color=self.style.wire_color,
+            )
+            self._ax.add_artist(wire_label)
+
+    def _draw_control_node(self, pos: int, xskip: float, color: str) -> None:
+        """
+        Draw the control node for the multi-qubit gate.
+
+        Parameters
+        ----------
+        pos : int
+            The position of the control node, in terms of the wire number.
+
+        xskip : float
+            The horizontal value for getting to requested layer.
+
+        color : str
+            The color of the control node. HEX code or color name supported by matplotlib are valid.
+        """
+
+        pos = pos + self._cwires
+
+        control_node = Circle(
+            (
+                xskip + self.style.gate_margin + self.style.gate_pad,
+                pos * self.style.wire_sep,
+            ),
+            self._control_node_r,
+            color=color,
+            zorder=self._zorder["node"],
+        )
+        self._ax.add_artist(control_node)
+
+    def _draw_target_node(self, pos: int, xskip: float, color: str) -> None:
```

Comment:
I understand that these target_node and control_node are for CNOT gate only?

Reply:
`_draw_control_node` is also utilized by some multi-qubit gates such as FREDKIN. Additionally,  if a user defines custom gates with a control node, `_draw_control_node` will be used. `_draw_target_node` is also used by the TOFFOLI gate.